### PR TITLE
Usage 3.0

### PR DESCRIPTION
**Pull request by devoncarew**
- fix a bug reporting the locale for the dart:io provider
- report additional stack trace lines for exceptions
- update some docs

@kevmoo

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 .packages
 .pub/
 pubspec.lock
+.vscode/
```
Comment (kevmoo) [inline .gitignore]: No! Global ignore, please.

File: example/ga.dart [hunk 2]
```diff
@@ -23,7 +23,8 @@ main(List args) async {
 
   await ga.sendScreenView('home');
   await ga.sendScreenView('files');
-  await ga.sendException('foo exception, line 123:56');
+  await ga
+      .sendException('foo error:\n' + sanitizeStacktrace(StackTrace.current));
```
Comment (kevmoo) [inline example/ga.dart]: Use String interpolation here, as is tradition.

File: lib/src/usage_impl.dart [hunk 3]
```diff
@@ -157,15 +155,21 @@ class AnalyticsImpl implements Analytics {
...
     // user's drive (file:/...).
     if (description.contains('file:/')) {
       description = description.substring(0, description.indexOf('file:/'));
     }
 
-    if (description != null && description.length > _MAX_EXCEPTION_LENGTH) {
-      description = description.substring(0, _MAX_EXCEPTION_LENGTH);
+    description = description.replaceAll('\n', '; ');
+
+    if (description != null && description.length > maxExceptionLength) {
```
Comment (kevmoo) [inline lib/src/usage_impl.dart]: this `null` check seems silly with the above usage, no?

File: lib/src/usage_impl_io.dart [hunk 4]
```diff
@@ -29,7 +32,12 @@ class AnalyticsIO extends AnalyticsImpl {
             new IOPostHandler(),
             applicationName: applicationName,
             applicationVersion: applicationVersion,
-            analyticsUrl: analyticsUrl);
+            analyticsUrl: analyticsUrl) {
+    final String locale = getPlatformLocale() ?? '';
+    if (locale != null) {
```
Comment (kevmoo) [inline lib/src/usage_impl_io.dart]: Won't be `null` here with the `??` above, correct?

File: lib/usage.dart [hunk 5]
```diff
@@ -297,10 +297,7 @@ String sanitizeStacktrace(dynamic st, {bool shorten: true}) {
 
   if (shorten) {
     // Shorten the stacktrace up a bit.
-    str = str
-        .replaceAll('(package:', '(')
-        .replaceAll('(dart:', '(')
-        .replaceAll(new RegExp(r'\s+'), ' ');
+    str = str.replaceAll(new RegExp(r'[\t ]+'), ' ');
```
Comment (kevmoo) [inline lib/usage.dart]: Could this be hit more than a few times? May want to make the `RegExp` an instance or top-level static var

File: lib/uuid/uuid.dart [hunk 6]
```diff
@@ -2,31 +2,25 @@
...
+///
+/// This will generate unique IDs in the format:
+///
+///     f47ac10b-58cc-4372-a567-0e02b2c3d479
+///
+/// The generated uuids are 128 bit numbers encoded in a specific string format.
+/// For more information, see
+/// [en.wikipedia.org/wiki/Universally_unique_identifier](http://en.wikipedia.org/wiki/Universally_unique_identifier).
 class Uuid {
   Random _random = new Random();
```
Comment (kevmoo) [inline lib/uuid/uuid.dart]: Should be final?

File: .gitignore (hunk 1 above)
Comment (devoncarew) [inline .gitignore]: done

File: example/ga.dart (hunk 2 above)
Comment (devoncarew) [inline example/ga.dart]: done

File: lib/src/usage_impl.dart (hunk 3 above)
Comment (devoncarew) [inline lib/src/usage_impl.dart]: yup, removed

File: lib/src/usage_impl_io.dart (hunk 4 above)
Comment (devoncarew) [inline lib/src/usage_impl_io.dart]: yup; remove the `??` operator - we only send the locale now if we can determine one

File: lib/usage.dart (hunk 5 above)
Comment (devoncarew) [inline lib/usage.dart]: makes sense - moved to a top-level final var

File: lib/uuid/uuid.dart (hunk 6 above)
Comment (devoncarew) [inline lib/uuid/uuid.dart]: done

**Pull request merged**